packages/core/src/common/connection.ts

```diff
@@ -0,0 +1,270 @@
+// *****************************************************************************
+// Copyright (C) 2022 Ericsson and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { injectable } from 'inversify';
+import { Disposable, DisposableCollection } from './disposable';
+import { Emitter, Event } from './event';
+import { MaybePromise, serviceIdentifier } from './types';
+
+export enum ConnectionState {
+    OPENING,
+    OPENED,
+    CLOSING,
+    CLOSED,
+}
+export namespace ConnectionState {
+    /**
+     * @param connection
+     * @returns Whether `connection` is or is going to be closed.
+     */
+    export function isClosed<T>(connection: Connection<T>): connection is Connection<T> & { state: ConnectionState.CLOSING | ConnectionState.CLOSED } {
+        return connection.state === ConnectionState.CLOSING || connection.state === ConnectionState.CLOSED;
+    }
+}
+
+/**
+ * Helper type to get a `Connection<any>` without triggering the linter.
+ */
+export type AnyConnection = Connection<any>;
+
+/**
+ * A `Connection` allows you to listen for messages and send messages back.
+ *
+ * Messages must arrive integrally and in the order they are sent!
+ *
+ * Most implementations are going to be API adapters.
+ */
+export const Connection = serviceIdentifier<Connection<any>>('Connection');
```

Comment:
Why is there a service identifier here?

Reply:
I used to bind named connections as part of internal wiring, having the identifier defined here helped.